### ISSUE TITLE
Update Result is always the same

### ISSUE THREAD

**Issue by omnitired** [typeorm/typeorm#2660]
**Issue type:**

[x] question
[ ] bug report
[] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[ ] `latest`
[x] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**
Hi.
the update result of the repository update is always the same even if update is unsuccessful.
so the code
```
let acc = await accountRepository.update({uuid: account.uuid}, {balance: price.total})
```
will run without rejecting even if the uuid is wrong. the result is always ```UpdateResult { generatedMaps: [], raw: [] }```. how can I throw an error if update didnt happen?

**Comment by maxxt** [comment]
Same case for me.

.update() will always return { generatedMaps: [] }

**Comment by giancarlo-dalle-mole** [comment]
Same for me on PostgreSQL 11.

Take the following example:

### Insert:

#### Entity:
```
Photo {
    name: 'Me and Bears',
    description: 'I am near polar bears',
    filename: 'photo-with-bears.jpg',
    views: 0,
    isPublished: true,
    price: '$0.99',
    date: '2018-11-19' 
    //createdAt -> Generated by database trigger on INSERT, annotated only with @Column
    //updatedAt -> Generate by database trigger on UPDATE, annotated only with @Column
}
```
#### Insert operation with QueryBuilder

``` TypeScript
const insertResult: InsertResult = await entityManager.createQueryBuilder().insert().values(photo).into(Photo).returning(["id", "createdAt", "updatedAt"]).updateEntity(true).execute();

//query: INSERT INTO "photos"("name", "price", "date", "description", "filename", "views", "isPublished", "created_at", "updated_at", "authorId") VALUES ($1, $2, $3, $4, $5, $6, $7, DEFAULT, DEFAULT, DEFAULT) RETURNING "id", "created_at", "updated_at", "price" -- PARAMETERS: ["Me and Bears","$0.99","2018-11-19","I am near polar bears","photo-with-bears.jpg",0,1]
```
#### Insert result and Entity
```
InsertResult {
    identifiers: [ { id: 46 } ],
    generatedMaps:
        [ {
            id: 46,
            createdAt: 2018-11-19T22:07:57.695Z,
            updatedAt: null,
            price: '$0.99' 
        } ],
    raw:
    [ {
        id: 46,
       created_at: 2018-11-19T22:07:57.695Z,
       updated_at: null,
       price: '$0.99' 
    } ]
 }

Photo {
    name: 'Me and Bears',
    description: 'I am near polar bears',
    filename: 'photo-with-bears.jpg',
    views: 0,
    isPublished: true,
    price: '$0.99',
    date: '2018-11-19',
    id: 46, // Returned by the database
    createdAt: 2018-11-19T22:07:57.695Z, // Returned by the database
    updatedAt: null // Returned by the database
}
```

The results are correct!.

### Update
```
Photo {
    name: 'Me and Bears',
    description: 'I am near polar bears',
    filename: 'photo-with-bears.jpg',
    views: 0,
    isPublished: true,
    price: '$0.99',
    date: '2018-11-19',
    id: 47, // Returned from the INSERT
    createdAt: 2018-11-19T22:20:44.878Z, // Returned from the INSERT
    updatedAt: null, // Returned from the INSERT, will be updated on UPDATE
    author: Author { name: 'Alice', id: 28 }  // Added an Author to the photo
}
```
#### Update operation (the ""wrong"" way)

```TypeScript
const updateResult: UpdateResult = await entityManager.createQueryBuilder().update(Photo, photo).where({id: photo.id}).returning(["id", "createdAt", "updatedAt"]).updateEntity(true).execute();

// query: UPDATE "photos" SET "name" = $2, "description" = $3, "filename" = $4, "views" = $5, "isPublished" = $6, "price" = $7, "date" = $8, "id" = $9, "created_at" = $10, "updated_at" = $11, "authorId" = $12 WHERE "id" = $1 RETURNING "id", "created_at", "updated_at" -- PARAMETERS: [47,"Me and Bears","I am near polar bears","photo-with-bears.jpg",0,1,"$0.99","2018-11-19",47,"2018-11-19T22:20:44.878Z",null,28]
```

#### Update result and Entity
```
UpdateResult {
    generatedMaps: [], // No generated maps!
    raw:
        [ { 
            id: 47,
           created_at: 2018-11-19T22:20:44.878Z,
           updated_at: 2018-11-19T22:20:45.085Z //The database generated a new timestamp!
      } ] 
}

Photo {
    name: 'Me and Bears',
    description: 'I am near polar bears',
    filename: 'photo-with-bears.jpg',
    views: 0,
    isPublished: true,
    price: '$0.99',
    date: '2018-11-19',
    id: 47,
    createdAt: 2018-11-19T22:20:44.878Z,
    updatedAt: null, // TypeORM did not update the field with the returning database value
    author: Author { name: 'Alice', id: 28 } }
```

This is not the behavior you expect and the documentation do not say anything about this behavior. But there is a way you can make things act the way we want.

### Update operation (the ""correct"" way)

```TypeScript
// We changed the "where" to use the "whereEntity", this tells TypeORM what are the entity being update and will have the values updated. 
// Again, the documentation do not mention this and no warning or exception is thrown.
const updateResult: UpdateResult = await entityManager.createQueryBuilder().update(Photo, photo).whereEntity(photo).returning(["id", "createdAt", "updatedAt"]).updateEntity(true).execute();

// query: UPDATE "photos" SET "name" = $2, "description" = $3, "filename" = $4, "views" = $5, "isPublished" = $6, "price" = $7, "date" = $8, "id" = $9, "created_at" = $10, "updated_at" = $11, "authorId" = $12 WHERE "id" = $1 RETURNING "id", "created_at", "updated_at" -- PARAMETERS: [48,"Me and Bears","I am near polar bears","photo-with-bears.jpg",0,1,"$0.99","2018-11-19",48,"2018-11-19T22:31:24.290Z",null,29]
```

#### Update result and Entity
```
UpdateResult {
    generatedMaps: // Now the Generated Maps are correctly generated
        [ { 
            id: 48,
           createdAt: 2018-11-19T22:31:24.290Z,
           updatedAt: 2018-11-19T22:31:24.494Z 
       } ],
    raw:
        [ {
            id: 48,
            created_at: 2018-11-19T22:31:24.290Z,
            updated_at: 2018-11-19T22:31:24.494Z 
        } ]
}

Photo {
    name: 'Me and Bears',
    description: 'I am near polar bears',
    filename: 'photo-with-bears.jpg',
    views: 0,
    isPublished: true,
    price: '$0.99',
    date: '2018-11-19',
    id: 48,
    createdAt: 2018-11-19T22:31:24.290Z,
    updatedAt: 2018-11-19T22:31:24.494Z, // The timestamp is set accordingly
    author: Author { name: 'Alice', id: 29 } 
}
```

As you can see, using the ```whereEntity``` in place of ```where``` when you have the entity (and the main reason you want the "generatedMaps"), is the way of doing updates and setting the values returned by the database on the entity.

An addendum: it would be nice to tell TypeORM which columns are generated on the database by triggers (or other mechanisms) depending on the operation. A flag on the ```@Column``` decorator would suffice. I could use the ```@CreateDateColumn``` and ```@UpdateDateColumn``` on the given example, but some times there are complex business logic running on the DB layer instead of the application layer (mainly on projects that use SQL Server or ORACLE), which makes maintenance easier on enterprise level projects.

**Comment by ywachao** [comment]
UpdateResult 每次返回的都是一样内容，怎么知道是更新成功了还是失败了？

**Comment by pleerock** [comment]
> the update result of the repository update is always the same even if update is unsuccessful.

its not right. If update updated zero rows it doesn't mean its unsuccessful - it means it just didn't find anything to update. We shouldn't throw any error in this case.

Currently data used in `UpdateResult` is mostly for internal usage, can you please point what exact data are you expecting to get in UpdateResult?

@giancarlo-dalle-mole some api is used internally, it doesn't suppose to be something that is supposed for user usage.

**Comment by billyjov** [comment]
@pleerock regarding this issue it may be usefull to get the updated data in the response e.g like this 

```json
{
 "title": "go to issue 2660",
 "name": "typeorm"
}

```


```json
 {
        "generatedMaps": [],
        "raw": {
            "fieldCount": 0,
            "affectedRows": 0,
            "insertId": 0,
            "serverStatus": 2,
            "warningCount": 0,
            "message": "(Rows matched: 0  Changed: 0  Warnings: 0",
            "protocol41": true,
            "changedRows": 0
        },
       "data": {
            "title": "go to issue 2660",
            "id": 91,
            "updatedAt": "2019-01-29T15:55:35.593Z"
         }
    }
```

**Comment by Kononnable** [comment]
Closing as no answer from the reporter.

**Comment by s97712** [comment]
How can I known either update success or update fail?

**Comment by DevelAx** [comment]
> How can I known either update success or update fail?
```TypeScript
const updateResult: UpdateResult = await repository.update(...);
```

Then look into `updateResult.raw.changedRows` or `updateResult.raw.affectedRows`.

**Comment by OLDIN** [comment]
I have the same problems. Is the problem still unsolved?

**Comment by s97712** [comment]
@OLDIN  I using queryBuilder replace of  repository.

**Comment by OLDIN** [comment]
@s97712 
I created a soft deletion method in repository.
```
public async softDeleteById(id: string) {
    const updateResult = await this.createQueryBuilder()
      .update()
      .set({
        [this.deletedAtColumnName]: () => 'NOW()'
      } as any)
      .where(`${this.deletedAtColumnName} IS NULL AND id = :id`, { id })
      .execute();

    return updateResult;
  }
```
And it still returns an empty result. Should I create a queryBuilder without touching the repository at all?

**Comment by s97712** [comment]
Try to add .returning() if you database support `returning`
```
public async softDeleteById(id: string) {
    const updateResult = await this.createQueryBuilder()
      .update()
      .set({
        [this.deletedAtColumnName]: () => 'NOW()'
      } as any)
      .where(`${this.deletedAtColumnName} IS NULL AND id = :id`, { id })
      .returning(["id"])
      .execute();

    return updateResult;
  }
```

**Comment by OLDIN** [comment]
@s97712 
generatedMaps is still empty, but the data is now in raw. 
```
UpdateResult {
    generatedMaps: [],
    raw: [ { id: '6cf16341-3562-46c6-a202-af76d9fb012c' } ]
}
```
At least some way to get the amount of updated data, thanks for the hint. I hope this wonderful library will soon work steadily.

**Comment by janober** [comment]
Btw did also just run into the same issue with sqlite: When "update()" gets used it always returns:
UpdateResult { generatedMaps: [], raw: [] }
No matter if something got updated or not always both arrays are empty.

**Comment by JannesV** [comment]
@Kononnable This needs to be re-opened. This is still an issue. (With MsSql for me)

**Comment by maxxims** [comment]
So this was acknowledged as being an issue and it still closed even though many users are still reporting the same issue. Nice job!

**Comment by Kononnable** [comment]
>Currently data used in UpdateResult is mostly for internal usage, can you please point what exact data are you expecting to get in UpdateResult?
@giancarlo-dalle-mole some api is used internally, it doesn't suppose to be something that is supposed for user usage.

I don't see an answer to this question, so I still think this issue can be closed as it it.

There is a small detail that people forget about open source - often there are more problems to solve/ideas to implement than we all have time to contribute. It leaves us with the fact that some ideas have to be forgotten because nobody have time to take care of them and if left open will eventually make such a mess that issue list will become unmaintainable.

If anyone thinks they know how this should be improved, please step forward and write few words about that. Write a proposal(probably in new issue - original issue here was about not throwing an error, which is correct behavior). Without a proposal it's not likely to be implemented in the near future - we have to prioritize our work which mostly mean fixing a bug or implementing something which already have an implementation proposal on which we all agree.

@maxxims 
Closing stale issues is normal thing in OSS and if no comment point us nearer to resolving the issue we can consider it stale. If questions were asked and nobody answered them it usually means no one cares about the issue anymore enough to spend even a little time on it.

**Comment by maxxims** [comment]
@Kononnable I appreciate the hard work you do and didn't mean to sound judgmental.

The thing is, this issue has at least 5 duplicates with many comments about users that can't find any way of knowing if an Update was successful or not.

The empty array returned says nothing about the Update operation performed.

**Comment by Kononnable** [comment]
@maxxims 
I understand your point of view. And I agree with it(at least partially).
The main problem(as I see it) is that we do not have enough firepower to cover all of the cases. For example we're developing two versions of typeorm for over a year now. It's normal to introduce breaking changes only in 'major' releases but the timeline of developing separate branches should be much shorter. However it requires some work to make such release(write migration guides etc.) and we didn't have time for it yet.
From my perspective(as a maintainer) the most pressing matters now are(in that order):
* releasing 0.3.0
* simplifying typeorm 'core' code(it would make code more maintainable and allow more community contributions)
* implementing new features which have proposals we already agreed on

There are also bugs to be fixed etc.

While I understand that the issue described here matters I know I won't have time to think how it should be implemented. That's why I'm encouraging you guys to make a concrete proposal. I can't force you to do so(and don't want to), but have to push you into that direction somehow.
Without a proposal I doubt it will be implemented in the next year even while it's pretty important issue and it would be great to get this resolved ASAP. I did not closed the issue because there is no problem here. I closed it to let everyone now that nothing will be done about it unless we have a proposal how to implement it. We have plenty of other issues with proposals, so this one would probably just sit here forever and from time to time it would get "+1" or similar comments which brings nothing.

**Comment by svstartuplab** [comment]
@Kononnable, I understand your constraints and appreciate your contributions.  I've donated a little money to the project and will try to donate more to help.  

I'm on the Nestjs, TypeORM, Postgres stack.  All my CRUD operations are working.  The issue is just error messages in the console with GraphQL.  I didn't have this issue with REST. It is annoying but it would be nice to get it fixed by returning whatever was deleted or updated.  It would save devs a lot of time chasing this error.  GraphQL requires the same shape object as in the schema.

I'm not competent enough for a PR, sorry.  What the code needs is to behave like .save with .update and .delete.  I suspect that you already have the needed code in .save.  It may be just a copy and paste level chore.  Maybe this will narrow the scope of the problem enough to get it fixed.

This returns the saved object, very good:
```
public async addItem(recordData): Promise<Members> {
    const result = await this.membersRepository.save(recordData);
    console.log('addItem res in service: ', result);
    return result;
```

Below are where the problems are:

```
public async deleteItem(recordId) {
    const result = await this.membersRepository.delete(recordId);
    console.log('delete res in service: ', result);
    return result;
  }
```
```
delete res in service:  DeleteResult { raw: [], affected: 1 }
```
```
public async updateItem(recordId, recordUpdate): Promise<any> {
    const result = await this.membersRepository.update(recordId, recordUpdate);
    console.log('update res in service: ', result);
    return result;
  }
```
```
update res in service:  UpdateResult { generatedMaps: [], raw: [] }
```

**Comment by developer239** [comment]
I understand that this is open source but I don't understand why is it closed. 🙂

**Comment by rustamwin** [comment]
As @Kononnable told

> Closing as no answer from the reporter.

**Comment by developer239** [comment]
For other people struggling with this my temporary solution is the following:

Simple delete higher order resolver:
https://github.com/developer239/node-type-orm-graphql/blob/master/src/modules/Core/resolvers/deleteResolver.ts

Simple update higher order resolver:
https://github.com/developer239/node-type-orm-graphql/blob/master/src/modules/Core/resolvers/updateResolver.ts

You can find the whole repo here:
https://github.com/developer239/node-type-orm-graphql

**Basically, I am doing 2 queries.** 😢

**Comment by CristianPumaES6** [comment]
es verdad lo que mencionan el update retorna lo mismo y no podemos asegurarnos si el update fue exitoso o ha fallado.

Para eso yo me aseguro antes de que todo este correcto.

``` typeorm

    // Actualiza un usuario
    async update(user: User): Promise<any> {
        // Hacemos una busqueda por id
        return await this.userRepository.findOne({
            where: [
                // hacemos un where donde buscamos por id.
                { id: user.id }
            ]
        }).then(resultFind => {
            // Validamos si encontro al usuario.
            if (!resultFind) throw new Error('user_does_not_exist');

            // verificamos que el email no este en uso, recordemos que el email es unico.
            return this.userRepository.findOne({
                where: [
                    // hacemos un where donde buscamos por nick o email y no sea del mismo id.
                    { id: Not(user.id) },
                    {
                        nick: user.nick,
                        email: user.email
                    }
                ]
            })
        }).then((result) => {
            // validamos si exite un resultado.
            if (result) {
                // Si existe, generamos un error para no continual.
                throw new Error('repeat_email_or_nick');
            }

            // encriptamos el password.
            return bcrypt.hash(user.password, 2);
        }).then(password => {
            // le asignamos el password encriptado al objeto
            user.password = password;

            // Actualizamos
            return this.userRepository.update(user.id, user);
        }).then(resultUpdate => {

            // borramos el password por seguridad.
            user.password = null;

            // Envio respuesta con el resultado recibido del ultimo paso
            return {
                status: true,
                message: (resultUpdate ? 'OK' : 'user update failed'),
                data: user
            };
        }).catch(
            err => {
                // Obtengo mensajes de error
                const clientMsg: string = (typeof err === 'string' ? err : 'CANNOT_PROCESS_REQUEST');
                const errorMsg: string = (typeof err === 'string' ? err : err.message || err.description || 'ERROR_EXEC_REQUEST');

                // Devuelvo respuesta con el mensaje obtenido
                return {
                    status: false,
                    error: errorMsg,
                    message: clientMsg,
                };
            }
        );
    }
```

**Comment by amirh715** [comment]
Any updates? I have the same problem. PLEASE!

**Comment by shellscape** [comment]
If this wasn't receiving replies on a regular basis, I could see the issue being marked and closed (and kept closed) as stale. But that's not the case. Pure laziness to sweep it under the rug. And I say that as a fervent Open Source contributor and maintainer on several large projects. Implement `stale-bot` to manage stale issues so users can keep issues open that need to be.

**Comment by amirh715** [comment]
Agreed! @shellscape I'm going to switch back to Sequelize. Too much unanswered questions and workarounds.

**Comment by jyotman** [comment]
The hack mentioned above by giancarlo-dm actually works for me too. I'm using Typescript and realized that using `whereEntity` instead of `where` and passing in an instance of Entity class instead of a regular object fixes it for me -
```
const result = await getUserRepository()
      .createQueryBuilder()
      .update(EntityModel, { employeeCount: 100 })
      .whereEntity({id: userId} as UserEntity) // Just need to cast to Entity Model instead of regular object
      .returning(['employeeCount', 'id'])
      .execute()
```

Result - 
```
UpdateResult {
  generatedMaps: [
    {
      id: 2,
      employeeCount: 100
    }
  ],
  raw: [
    {
      id: 2,
      employee_count
    }
  ],
  affected: 1
}
```

**Comment by godtaehee** [comment]
It's solved at database which support `returning` and `output`

But, In my case, I use SQLite3. So I can't get any solution in this Issue

Can I help with some solution in SQLite3??

**Comment by Yuvraj102** [comment]
i have got some similar issue couldnt figure out yet https://stackoverflow.com/questions/74068602/typeorm-update-method-doesnt-work-when-using-repository

**Comment by KMJ-007** [comment]
i have explained the issue and solution in gist 

https://gist.github.com/KMJ-007/1e58788f31f772bb422e5ebc8195f2a4,

i need some guidance to complete the PR
